chain/chaintest/block.go

```diff
@@ -104,3 +110,265 @@ func GenerateTestExecutedBlocks(
+				return nil, err
+			}
+			tsv := ts.NewView(stateKeys, parentCtx.view, 0)
+			if err := tx.PreExecute(ctx, feeManager, balanceHandler, rules, tsv, timestamp); err != nil {
+				return nil, err
+			}
+			result, err := tx.Execute(ctx, feeManager, balanceHandler, rules, tsv, timestamp)
+			if err != nil {
+				return nil, err
+			}
+			ok, d := feeManager.Consume(result.Units, rules.GetMaxBlockUnits())
+			if !ok {
+				return nil, fmt.Errorf("consumed too many units in dim %d", d)
+			}
+			tsv.Commit()
+		}
+
+		tsv := ts.NewView(state.CompletePermissions, parentCtx.view, 0)
+		if err := tsv.Insert(ctx, chain.HeightKey(metadataManager.HeightPrefix()), binary.BigEndian.AppendUint64(nil, height)); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.TimestampKey(metadataManager.TimestampPrefix()), binary.BigEndian.AppendUint64(nil, uint64(timestamp))); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.FeeKey(metadataManager.FeePrefix()), feeManager.Bytes()); err != nil {
+			return nil, err
+		}
+		tsv.Commit()
+
+		parentRoot, err := parentCtx.view.GetMerkleRoot(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		parentCtx.view, err = parentCtx.view.NewView(ctx, merkledb.ViewChanges{
+			MapOps:       ts.ChangedKeys(),
+			ConsumeBytes: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		statelessBlock, err := chain.NewStatelessBlock(
+			parentCtx.id,
+			timestamp,
+			height,
+			txs,
+			parentRoot,
+			nil,
+		)
+		if err != nil {
+			return nil, err
+		}
+		parentCtx.id = statelessBlock.GetID()
+
+		blk := chain.NewExecutionBlock(statelessBlock)
+		executionBlocks[i] = blk
+	}
+	return executionBlocks, nil
+}
+
+// BlockBenchmark is a parameterized benchmark. It generates NumBlocks
+// with NumTxsPerBlock, and then calls Processor.Execute to process the block
+// list b.N times.
+type BlockBenchmark struct {
+	MetadataManager chain.MetadataManager
+	BalanceHandler  chain.BalanceHandler
+	RuleFactory     chain.RuleFactory
+	AuthEngines     chain.AuthEngines
+
+	Config                chain.Config
+	AuthVerificationCores int
+
+	BlockBenchmarkHelper BlockBenchmarkHelper
+
+	// NumBlocks is set as a hyperparameter to avoid using b.N to generate
+	// the number of blocks to run the benchmark on.
+	NumBlocks      uint64
+	NumTxsPerBlock uint64
+}
+
+func (test *BlockBenchmark) Run(ctx context.Context, b *testing.B) {
+	r := require.New(b)
+
+	metrics, err := chain.NewMetrics(prometheus.NewRegistry())
+	r.NoError(err)
+
+	db, err := merkledb.New(
+		ctx,
+		memdb.New(),
+		merkledb.Config{
+			BranchFactor: merkledb.BranchFactor16,
+			Tracer:       trace.Noop,
+		},
+	)
+	r.NoError(err)
+
+	numAuthVerificationJobs := 100
+
+	processorWorkers := workers.NewSerial()
+	if test.AuthVerificationCores > 1 {
+		processorWorkers = workers.NewParallel(test.AuthVerificationCores, numAuthVerificationJobs)
+	}
+
+	chainIndex := &validitywindowtest.MockChainIndex[*chain.Transaction]{}
```

Comment:
This is fine because we never need to fetch a processing block during this execution benchmark.